### ISSUE TITLE
Feature request: Support multiple working tree

### ISSUE THREAD

**Issue by ppggff** [ForkIssues/Tracker#527]
Support function as `git worktree` that manage multiple working trees attached to the same repository.

**Comment by DanPristupov** [comment]
Is there any public available repository with work trees, so I could see how this feature is used in a real project?

**Comment by ppggff** [comment]
This feature is mainly for local development. I use this function to work with different branches at the same time, each working tree for a different branch. Then I don't need to checkout to different branch every time, saving a lot of time for rebuilding and other stuff.

**Comment by hardkrash** [comment]
Agreed,  worktree is a great feature.

It is entirely local, not something that would be seen on a git server.

To try it out on this project

git worktree add ../fork_issues_worktree 859ddf3c7663bed8cf913f761f1063dda56e9e89

Will make a worktree in your parent directory pointing to that commit.

**Comment by k3n** [comment]
Yeah, was looking forward to using this app, but without worktree support it does me no good -- it doesn't find any of the worktree repos.

**Comment by eugioc** [comment]
+1 for feature request for support for worktrees. It's such a useful git feature.

**Comment by natronite** [comment]
Every time I rescan a folder for repositories it removes the ones I added through `git worktree`. The thing is that the worktree folders don't contain a `.git` folder but a `.git` file.

**Comment by mariusmarais** [comment]
My team use worktrees heavily for local development, I'm putting my usage here for reference and comment by others.

I clone projects into folders as follows (but there is no actual structure requirement in git itself):
```
~/projects
  - project1
    - repo (usually master)
    - feature1 (marius/feature1)
    - featute2 (marius/feature2)
  - project2
    - repo (usually master)
    - feature1 (marius/feature1)
```

While this works, Fork sees every directory as a different repo, which confuses things. However, it makes the Fork recent items almost useless, since everything is called "repo" or "feature1".  Usually with a frontend and backend project on feature branches with the same name, it gets confusing, which is why I'm mostly using Terminal to launch Fork from the directories.

I think worktree support along the following lines would be quite useful:
 - If usage of worktrees is detected, determine the project name in a different way, by maybe looking at the parent directory or using a custom alias in .git/config?
 - For each worktree, show it somewhere under the project browser, BUT
    - Use the same interface window for all worktrees since they have the same history (this works well right now)
    - Under "Local Changes", list separate worktrees, with their own local changes underneath.
That would allow worktree navigation to be a lot more intuitive.

Actually creating new worktrees and so on would be nice, but this happens a lot less often than browsing, committing and so on, so I don't think that would be necessary for the main interface. Especially considering that anyone using worktrees has already figured out how to create one :) 

Using worktrees one is also less likely to switch a specific worktree to a different branch, you simply create a new one.

**Comment by laurentvaills** [comment]
I also use worktrees locally to checkout some branches besides my main directory. 
But unfortunately when I do so, I can't use Fork on this new directories and have to go back (and remember) the git command line.  
The "add interactive" is much easier through Fork rather than the command-line. 

If also I could select a branch in Fork and select "Checkout in an other directory" or so that would be really nice.

**Comment by mariusmarais** [comment]
@laurentvaills you can actually use Fork to stage & commit, I do so every day...

Just:
```sh
$ cd ~/worktree
$ fork .
```

And Fork will open it as a "new" repository, sharing history with the main repo.  You can have multiple worktrees open at the same time like this.  The repo browser window gets a bit confused but it works well enough if you always (mostly) use the CLI to launch.

**Comment by mariusmarais** [comment]
Explained worktrees to some new colleagues recently, and came to check up on this issue...

Separating out a folder for each worktree under "Local Changes" would really make a world of difference.

**Comment by DanPristupov** [comment]
@mariusmarais 

> Separating out a folder for each worktree under "Local Changes" would really make a world of difference.

The active branch and commit also differ. I think you should open both worktrees in separate tabs.

**Comment by mariusmarais** [comment]
Yes @DanPristupov, you're right. They could be shown as a different type of label in the history, and could be shown in the "folder" as a label, since that's where they apply.

But this is why "just" doesn't belong in feature requests, because it's never quite that simple.

**Comment by infotexture** [comment]
ℹ️ For details, see https://git-scm.com/docs/git-worktree.

I use worktrees all the time and find them very handy. While Fork doesn't detect worktrees from the main repo, it works fine if you switch to the worktree directory via the CLI as @mariusmarais [suggested](https://github.com/fork-dev/Tracker/issues/527#issuecomment-732303315) (or via **File > Open**, or **File > Open Recent**).

In this case, Fork shows all the parent repo's branches, tags, and stashes in the worktree repo window, so you can see everything in one place.

Would be nice to teach Fork how to “see” related worktrees from a parent repo and represent them in the sidebar somehow.

@DanPristupov The UI support could build on the approach for submodules, so there'd be an extra **Worktrees** section in the sidebar (below Tags, Stashes, & Submodules), with commands to add/delete worktrees, and maybe warn you if you delete a branch that's checked out in a worktree.

Double-clicking a linked worktree in the sidebar could open the worktree folder in a new Fork window, just like with submodules:

<details>
<summary><b>Worktrees</b></summary>
<ul>
<li>🔗 feature1
<li>🔗 feature2
</details>

**Comment by kikeenrique** [comment]
I would love to see some support for worktree in fork.

I've done some research to find git UIs supporting worktree and there are not many. I've found:
* gitlens, with a small preview in https://www.gitkraken.com/gitlens/plus-features 
<img width="1347" alt="gitlens worktree preview" src="https://user-images.githubusercontent.com/2754671/225900729-1ef77f03-ec71-439c-9c7f-22809a146171.png">

* tortoisegit, without a preview https://tortoisegit.org/docs/tortoisegit/tgit-dug-worktrees.html

I like the approach comment by @infotexture, having another sidebar like submodules.

**Comment by quincunx** [comment]
> I would love to see some support for worktree in fork.
> 
> I've done some research to find git UIs supporting worktree and there are not many. I've found:
> 
> * gitlens, with a small preview in https://www.gitkraken.com/gitlens/plus-features
> 
> <img alt="gitlens worktree preview" width="1347" src="https://user-images.githubusercontent.com/2754671/225900729-1ef77f03-ec71-439c-9c7f-22809a146171.png">
> 
> * tortoisegit, without a preview https://tortoisegit.org/docs/tortoisegit/tgit-dug-worktrees.html
> 
> I like the approach comment by @infotexture, having another sidebar like submodules.

Another great Git GUI with worktree support is:

* [SmartGit](https://www.syntevo.com/smartgit/)

**Comment by kikeenrique** [comment]
After using git worktree for some weeks and using fork as UI I've detected this problems:
* Fork does not see other worktrees (as this feature request asks)
* In order to open a worktree it's needed to go to the folder where the worktree is located and via CLI launch fork from that directory.
* "Show in finder" does not work for worktrees.

**Comment by DanPristupov** [comment]
@kikeenrique 

> In order to open a worktree it's needed to go to the folder where the worktree is located and via CLI launch fork from that directory.

You can use Cmd+O as an alternative.

> "Show in finder" does not work for worktrees.

It works for me. Check the log file for clues `~/Library/Logs/Fork.log` or send it to support@fork.dev.

**Comment by kikeenrique** [comment]
> @kikeenrique
> 
> > In order to open a worktree it's needed to go to the folder where the worktree is located and via CLI launch fork from that directory.
> 
> You can use Cmd+O as an alternative.

Thanks, I forgot that.
> 
> > "Show in finder" does not work for worktrees.
> 
> It works for me. Check the log file for clues `~/Library/Logs/Fork.log` or send it to [support@fork.dev](mailto:support@fork.dev).

Thanks for the log clue. 
I didn't see the problem clearly, in some worktrees it worked properly, but in others didn't. I've found a fix, apparently, which consists in closing the not working tab and opening again (from CLI, when opening it from cmd+O, "show in finder" **does not work**).
I can only see one log line with a failure, but it seems to be not related.
`❌[GetReferencesGitCommand.swift:89] > Cannot parse reference refs/original/refs/heads/fastlaneSwift

**Comment by DanPristupov** [comment]
@kikeenrique
> from CLI, when opening it from cmd+O, "show in finder" does not work

Thank you, I'll check this.

We will also have a look if we can show workspaces on the sidebar.

**Comment by brokenthorn** [comment]
> @mariusmarais
> 
> > Separating out a folder for each worktree under "Local Changes" would really make a world of difference.
> 
> The active branch and commit also differ. I think you should open both worktrees in separate tabs.

Work trees are basically separate checkouts of separate branches of the same repo, if I understood correctly.

What Fork I think needs to do is support them under the hood so it knows the edge cases that apply when issuing the usual git operations in a worktree versus a normal repo.

Visually, the functionality should be the same as with a normal repo. Perhaps it would be great if it could identify all worktrees of the same repo and perhaps add new roots to the top left "local changes" and "all commits" navigator.

![image](https://github.com/fork-dev/Tracker/assets/2119004/732f5cfa-7996-4633-a076-c2781ac7ee66)

to something like

![image](https://github.com/fork-dev/Tracker/assets/2119004/0adb07d8-eab3-46a0-8e9a-60fd73525361)

HTH.

I've just stumbled across worktrees and was thinking about using them and came here to see if Fork supports them, so:
I don't know if the repo name can still be identified, if not then probably just show worktree names.
I don't know if worktrees can be identified either, because you can technically spawn them wherever.
I don't know if worktrees are allowed to have branches.

**Comment by DanPristupov** [comment]
As an experiment, in the next update (Fork 2.31) we are going to add the list of existing worktrees to the sidebar:
<img width="281" alt="Screenshot 2023-07-25 at 17 07 16" src="https://github.com/fork-dev/Tracker/assets/618115/a95a211a-cd84-48a1-9b30-8ade2c8e938f">

**Comment by krin-san** [comment]
> @kikeenrique
> 
> > from CLI, when opening it from cmd+O, "show in finder" does not work
> 
> Thank you, I'll check this.
> 
> We will also have a look if we can show workspaces on the sidebar.

I have the same issue in Fork Version 2.33 with worktrees opened via CLI command `fork`. I have set up a "Branch Command" to automate creating worktrees which looks like that:
```sh
NAME="${repo:name}-$(echo ${ref:short} | tr '/' '\n' | tail -1)"
git worktree add -B "${ref:short}" ../"$NAME" ${ref}
fork ../"$NAME"
```
In the repo tab opened in Fork, "Open in Terminal" starts a shell correctly, in the worktree directory, but "Open in Finder" opens the main worktree directory instead – the one containing `.git` dir.

**Comment by brokenthorn** [comment]
> As an experiment, in the next update (Fork 2.31) we are going to add the list of existing worktrees to the sidebar: <img alt="Screenshot 2023-07-25 at 17 07 16" width="281" src="https://user-images.githubusercontent.com/618115/255945200-a95a211a-cd84-48a1-9b30-8ade2c8e938f.png">

This is super awesome to hear, Dan! So very excited for this feature!

May I add something? Hopefully this is useful feedback. Worktrees are conceptually somewhat like checking out a branch inside a dedicated folder instead of in the main worktree, where you can only have one branch checked out at a time. This allows you check out multiple branches to separate folders and switch between more easily, without having to commit or stash WIP work.

Thus, to me they look like just branches, conceptually. So, I would really love to have worktrees live closer somehow to the branches on the sidebar. Either next to Branches or if you can find a way to integrated them somehow... like Branches to contain Worktrees? IDK?

Anyway, I'm sure whatever you do, it's gonna be awesome!

**Comment by andymac4182** [comment]
Just used this! So excited to see this coming :) I love worktrees and its one of the few things I still have to do in the CLI.

**Comment by DanPristupov** [comment]
@brokenthorn 

Thank you for the feedback.

> ... to me [workspaces] look like just [branches], conceptually. So, I would really love to have `Worktrees` live closer somehow to `Branches` on the sidebar.

Conceptually the `Workspaces` section should be even **above** `Branches` because the worktrees concept is lying between working directory and branches. So, may be it should be the first item on the sidebar, right below the filter box.

Would that make the life better, though? It's a big change and I'm afraid of making something very wrong :).

@mtschoen-unity it would be great if you could share your point of view.

**Comment by brokenthorn** [comment]
> @brokenthorn 
> 
> Thank you for the feedback.
> 
> > ... to me [workspaces] look like just [branches], conceptually. So, I would really love to have `Worktrees` live closer somehow to `Branches` on the sidebar.
> 
> Conceptually the `Workspaces` section should be even **above** `Branches` because the worktrees concept is lying between working directory and branches. So, may be it should be the first item on the sidebar, right below the filter box.

I think you might be right... Let's hear more feedback though.

> Would that make the life better, though? It's a big change and I'm afraid of making something very wrong :).

Yeah. It's a higher level abstraction

Repo name (virtual) / n workspaces (physical, in different folders / branches / commits

Is the above the correct hierarchy model?

**Comment by infotexture** [comment]
> Conceptually the `Workspaces` section should be even **above** `Branches` because the worktrees concept is lying between working directory and branches. So, may be it should be the first item on the sidebar, right below the filter box.

That's true, but workspaces are a niche feature that even many Git experts don't use, so I'm not sure placing it first would make everyone happy. 

It would probably be fine if it remains context-sensitive, so the the `Workspaces` section only appears if additional workspaces are defined (like it is now).

Of course, you could always allow users to change the order of sidebar sections (like in Finder, where you can drag and drop sections into the preferred order), but that's probably best treated as a separate issue.

**Comment by chucker** [comment]
> Conceptually the Workspaces section should be even above Branches because the worktrees concept is lying between working directory and branches

Agreed. In terms of taxonomy, that makes sense to me. And since it only shows up if there _are_ additional worktrees, it shouldn't confuse existing users much.

Having said that: the flipside is that this behavior (it showing up dynamically) seems to be unlike any other module. For example, submodules are also a bit rarely used, yet they always show up. And that's useful, because it's how you _add_ submodules as well. (Whereas worktrees currently seem to be read-only in Fork.)

**Comment by brokenthorn** [comment]
> @brokenthorn
> 
> Thank you for the feedback.
> 
> > ... to me [workspaces] look like just [branches], conceptually. So, I would really love to have `Worktrees` live closer somehow to `Branches` on the sidebar.
> 
> Conceptually the `Workspaces` section should be even **above** `Branches` because the worktrees concept is lying between working directory and branches. So, may be it should be the first item on the sidebar, right below the filter box.
> 
> Would that make the life better, though? It's a big change and I'm afraid of making something very wrong :).
> 
> @mtschoen-unity it would be great if you could share your point of view.

Sorry. I meant work trees, not workspaces. I don't use the workspaces feature in Fork (yet) at all. So I've always referred to Git work trees in this discussion, not Fork workspaces. Sorry for the confusion.

What you are saying about workspaces being above branches is totally correct.
And yes, work trees are a concept between the working directory and branches.

Git has a concept of the default working directory, which is where you normally first check out your remote repository or where you initially init a git repo.

When you create a new git work tree, you create a folder that is akin to cloning your repository's main work tree, or akin to checking out (cloning) the same repository but in another separate folder, such that you can have different branches checked out in each of these folders, giving you the ability to have two "working trees" of the same repository, at the same time, and the ability for them to both have uncommitted changes, which is not something you can do with a single working tree (like the default one for example). You have to first discard your changes or commit them before you can switch to another branch.

So yeah, work trees are sort of in between: REPO / WORK DIR/TREE / ACTIVE BRANCH.

It's almost 4 AM here, so sorry if I made any mistakes. Thanks for listening!

**Comment by mtschoen-unity** [comment]
> @mtschoen-unity it would be great if you could share your point of view.

Hey there! I did a skim of the thread but I want to make sure we're talking about the same thing. Is the question about where to put the new Worktrees section in this ui?

<img width="251" alt="Screenshot 2023-10-16 at 2 19 28 PM" src="https://github.com/fork-dev/Tracker/assets/19391263/d729cb82-90f6-4965-8ba6-8615c1d3a834">

To be honest, I don't have much of a preference here. I don't tend to use that section at all, but instead I start with the worktree's representation in the `Repositories` or `Recent` list in the Repository Manager window. Staying at the bottom of the list would actually best fit my workflow. 🤷

**Comment by matheusamazonas** [comment]
Is it normal that, when a worktree is open, the remote branches are never displayed, neither on the sidebar nor on the commit history?

Edit: running `git remote show origin` does show all branches under the `Local refs configured for 'git push'` section.

**Comment by DanPristupov** [comment]
> Is it normal that, when a worktree is open, the remote branches are never displayed, neither on the sidebar nor on the commit history?

Yes, if it's a working tree for a bare repo. Otherwise it's not.

**Comment by matheusamazonas** [comment]
> Yes, if it's a working tree for a bare repo. Otherwise it's not.

That was the case. I just tested it with a non-bare repo and the remotes are there.

Just curious: is that a git or Fork limitation?

**Comment by DanPristupov** [comment]
> Just curious: is that a git or Fork limitation?

It's git. I think bare repos are supposed to be used on servers and can't have other remotes.

**Comment by matheusamazonas** [comment]
For those who are interested in using bare repositories, this [workaround](https://morgan.cugerone.com/blog/workarounds-to-git-worktree-using-bare-repository-and-cannot-fetch-remote-branches/) that adds remote tracking and that works with Fork too.

**Comment by DanPristupov** [comment]
In Fork 2.41 we added the ability to create new workspaces from a local branch.

<img width="500" alt="Screenshot 2024-03-22 at 17 30 23" src="https://github.com/fork-dev/Tracker/assets/618115/8cb1a688-781c-4eba-986b-a0fd53786b77">

**Comment by krin-san** [comment]
Nice to see more and more features related to Git Worktrees. I have a couple of custom Branch and Repository scripts to create and delete worktrees making it an ideal feature to work on multiple tasks in parallel.

I kinda expected to find this 2.41 feature in a Branch context menu. It is an ideal place for such a feature because it is contextual – you already clicked on the branch you want to create a worktree for, but have to go the main menu instead and search this branch name again.

What do you think about adding "Add Worktree..." to the Branch context menu as well?
<img width="732" alt="Screenshot 2024-03-26 at 09 24 35" src="https://github.com/fork-dev/Tracker/assets/2318002/d5af698e-5364-4d31-85d5-0cdac5685197">

**Comment by DanPristupov** [comment]
I'd like to share this post: https://matklad.github.io/2024/07/25/git-worktrees.html. I think worktrees users might find it interesting.

**Comment by ghost** [comment]
> It's git. I think bare repos are supposed to be used on servers and can't have other remotes.

I am using bare-repo backed work-trees and remote operations work fine from CLI, but not from Fork.

My test just now:
- Made a new bare clone of one of my upstream repo's from GitHub
    `git clone --bare git@github.com:me/myRepo.git`
- Added a worktree from it:
    `git --git-dir=./myRepo.git worktree add myRepo-develop develop`
- ✅  Successfully performed remote `fetch` / `pull` / `push` to `origin` from within `myRepo-develop` in the CLI.
- ℹ No [editing of config](https://morgan.cugerone.com/blog/workarounds-to-git-worktree-using-bare-repository-and-cannot-fetch-remote-branches/) in `.git` was required (@matheusamazonas )

Perhaps Git support has improved? My `git --version` reports `git version 2.39.5 (Apple Git-154)`

@matheusamazonas @DanPristupov This result seems misaligned with your assertions above.

**Comment by matheusamazonas** [comment]
> This result seems misaligned with your assertions above.

My comment is from over a year ago, so it might not be relevant anymore. As you said, maybe git changed ever since.

Regarding Fork, as @DanPristupov said, support for adding worktrees was added on 2.41. But in my experience, making it work with bare repos needs one extra step. Here's what I do:
- Close the repo with `--bare`.
- Create a new worktree with `git worktree`, using a branch that is not the one you'd like to use (it will be a throwaway worktree, just for setting it up).
- Open that worktree on Fork. It loads successfully, but the remote data (e.g. branches) doesn't show up.
- Using Fork, create a new worktree, on the branch you'd like to use. This worktree loads successfully, including all its remote data and branches. Success!
- Delete the first worktree (the throwaway one).
- Profit.

It's not ideal, but it gets the job done.

**Comment by dineshm72** [comment]
> I'd like to share this post: https://matklad.github.io/2024/07/25/git-worktrees.html. I think worktrees users might find it interesting.

Interesting.  I do agree that stashing can be dangerous/complicated, and committing everything every time you want to switch branches is a workable process if you don't want to stash and branch.  But to me worktrees are not at all replacements for branches.  They are replacements for having to clone multiple copies of the same repo, duplicating all the git backend data stores, in order to simultaneously have multiple branches "exist" in the filesystem to allow for multiple builds or testing environments.

The current implementation of worktrees in fork is helpful but only half implemented to me.  Features I miss:

* There is no way in the UI to see what repos are primary and which are worktrees and how they relate (i.e. git worktree list)
* There is no way to remove or prune worktrees in fork, you have to do it all in CLI
* There is no way to open a worktree that you didn't just create with "Add Worktree...", e.g. if you want to rename a worktree directory in the filesystem, it becomes disconnected (just like a normal repo would), but there is no way to get it back in fork (trying to open it throws an error), you have to just remove and prune it from CLI and then recreate it, saving any in process work manually if you want to keep it

**Comment by kikeenrique** [comment]
List, Add and Delete worktrees is currently working fine for me.

https://github.com/user-attachments/assets/21b858b1-fe50-4561-95eb-592b2e6d0805

In fact, this PR shall be closed by now. @DanPristupov

**Comment by dineshm72** [comment]
Very strange, I saw a reference to that Worktree sidebar element and when I looked there wasn't even a Worktree heading, I thought it might have been a Windows only feature.  But it's there now, maybe it was just a transient (or user) issue on my end.

Agreed, now that it's there everything seems to work very well, it seems to very quickly update the state of the worktree....tree if the filesystem changes underneath it (both if the expected directory disappears and if it is restored).  There is no access to the repair command, but honestly that is something that can be done in CLI only in the very infrequent case it is needed rather than building out a whole UI for it.  

Great news, thanks for pointing me to it!

**Comment by HansFalkenberg-Visma** [comment]
Thanks for making this mostly work! I am hoping you can make it work for WSL too though.

WSL repos do not show worktrees. It shows the `Worktrees` category (below Submodules), and the repository name. But with none of the worktrees below it.

Maybe because the `.git\worktrees\<tree-name>\gitdir` files (and the `git worktree list` command) use the WSL local path and this does not at all match the repository path as Fork sees it? Which is something like `\\wsl.localhost\Ubuntu-24.04\<local-path-except-with-backslashes>`

**Comment by AcrylicShrimp** [comment]
I cannot see the `Worktrees` element in the left panel, too.

<img width="184" height="271" alt="Image" src="https://github.com/user-attachments/assets/9ea28ab1-cd01-4ee1-bb44-f89a28709077" />

**Comment by DanPristupov** [comment]
@AcrylicShrimp run `git worktree list` in the command line and send the result to support@fork.dev

**Comment by tyteen4a03** [comment]
Is it possible to create worktrees directly from Fork right now?

**Comment by dineshm72** [comment]
Yes, you can create a worktree from a tracked branch (that you do not currently have checked out) by right clicking on it in the left sidebar:

<img width="570" height="209" alt="Image" src="https://github.com/user-attachments/assets/5cf3dd48-9a0a-4d55-a43b-04e63f507b6e" />

Or you can create a new branch from an existing branch and make it a worktree from the Repo menu item:

<img width="431" height="356" alt="Image" src="https://github.com/user-attachments/assets/d32eca55-07ff-461b-a4f7-be0a55489923" />

Both are windows client, but I am pretty sure they work the same way in mac client.

**Comment by tyteen4a03** [comment]
@dineshm72 that doesn't work for branches that has not been fetched yet (i.e. only on remote)

**Comment by dineshm72** [comment]
I have never tested that, but I guess you're right.  Is that an issue for you?

You need to create a local tracking branch first in Fork it seems, which appears to be a limitation that doesn't exist in git.  On the other hand, doing it based on a remote branch directly (in git) puts you into a detached head state on the worktree, which I usually try to avoid anyway.

Unless the dev decides to enhance the worktree support in Fork I guess you'd just need to do it in git if you need that functionality.

**Comment by tyteen4a03** [comment]
@dineshm72 Yes, as I am increasingly using worktrees for AI coding agents. Harnesses such as T3 code can do this natively.

**Comment by dineshm72** [comment]
OK, that makes perfect sense.  But then presumably you don't need Fork at all, the agents can create the worktree using git in the terminal.  Once the worktree is created I believe Fork should pick it up automatically so you can view it in Fork if you want to.

**Comment by tyteen4a03** [comment]
I still like using Fork to view the diff myself and see the entire Git tree, which isn't something coding agent harnesses are good at at the moment

**Comment by dineshm72** [comment]
As I said, I think Fork will pick up the worktree if the agent creates it in the terminal.  But you might want to test to be sure.
